server/src/client_api.rs

```diff
@@ -65,44 +68,39 @@ async fn resolve_features(
         .get(&edge_token.token)
         .map(|e| e.value().clone())
         .ok_or(EdgeError::AuthorizationDenied)?;
-    let client_features = match req.app_data::<Data<FeatureRefresher>>() {
-        Some(refresher) => refresher.features_for_token(validated_token.clone()).await,
-        None => features_cache
-            .get(&cache_key(&edge_token))
-            .map(|features| features.value().clone())
-            .map(|client_features| ClientFeatures {
-                features: client_features
-                    .features
-                    .filter_by_projects(&validated_token),
-                ..client_features
-            })
-            .ok_or(EdgeError::ClientFeaturesFetchError(FeatureError::Retriable)),
-    };
-    let filters = filter_query.into_inner();
+
+    let query_filters = filter_query.into_inner();
     let query = unleash_types::client_features::Query {
         tags: None,
-        projects: Some(validated_token.projects),
-        name_prefix: filters.name_prefix.clone(),
-        environment: validated_token.environment,
+        projects: Some(validated_token.projects.clone()),
+        name_prefix: query_filters.name_prefix.clone(),
+        environment: validated_token.environment.clone(),
         inline_segment_constraints: Some(false),
     };
-    client_features
-        .map(|f| ClientFeatures {
-            query: Some(query),
-            features: f
-                .features
-                .into_iter()
-                .filter(|f| {
-                    filters
-                        .name_prefix
-                        .as_ref()
-                        .map(|prefix| f.name.starts_with(prefix))
-                        .unwrap_or(true)
-                })
-                .collect(),
-            ..f
-        })
-        .map(Json)
+
+    let filter_set = if let Some(name_prefix) = query_filters.name_prefix {
+        FeatureFilterSet::from(Box::new(name_prefix_filter(name_prefix)))
+    } else {
+        FeatureFilterSet::default()
+    }
+    .with_filter(project_filter(&edge_token));
+
+    let client_features = match req.app_data::<Data<FeatureRefresher>>() {
+        Some(refresher) => {
+            refresher
+                .features_for_filter(validated_token.clone(), filter_set)
+                .await
```

Comment:
I feel that this call to the refresher should be equal to calling `features_cache.get(&cache_key(&edge_token))` (because it would populate the cache on the specific same key). Then mapping and filtering should happen independently of the path we took to get the client_features.

Something like:
```
if (I have to refresh) 
   client_features = get from the refresher
else 
   client_features = get from cache
client_features.map(|client_features| filter_client_features(&client_features, filter_set))
            .ok_or(EdgeError::ClientFeaturesFetchError(FeatureError::Retriable))
```

So I'm a bit confused because the refresher also accesses a features_cache (maybe a different one?)

Reply:
Yeah the code is a little confusing here, this needs a little love in the near future perhaps. What's happening here is that the `feature_refresher` holds its own internal cache in `edge` mode and if you're in `offline` mode you use the other cache.

The filtering on demand is a lot harder to do with our architecture, unfortunately, this is us not being kind to Rust's memory model, but ultimately, it's hard to return references to the `client_features` without massively violating the lifetime rules. I would like to get to your suggestion but without a large overhaul it's tough.

Needs some refactor in a future body of work I think